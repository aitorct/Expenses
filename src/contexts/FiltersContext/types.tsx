export interface FiltersContextInterface {
  filters: Filters;
  applyFilters: (key: FilterType, value: FilterValue) => void;
  filtersActive: boolean;
}

export enum FilterType {
  USER = 'user',
  YEAR = 'year',
  RECEIPTS = 'receipts',
}

export interface Filter {
  value: FilterValue;
}

export type FilterValue = string | null;

export type Filters = Record<FilterType, Filter>;
