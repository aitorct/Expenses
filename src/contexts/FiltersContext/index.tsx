import React, {ReactNode, useState} from 'react';
import {
  Filters,
  FiltersContextInterface,
  FilterType,
  FilterValue,
} from './types';

const defaultFilters = {
  user: {
    value: null,
  },
  year: {
    value: null,
  },
  receipts: {
    value: null,
  },
};

export const FiltersContext =
  React.createContext<FiltersContextInterface | null>(null);

export const FiltersProvider = ({children}: {children: ReactNode}) => {
  const [filters, setFilters] = useState<Filters>(defaultFilters);

  const applyFilters = (key: FilterType, value: FilterValue) => {
    const updatedFilters = {...filters, [key]: {value}};
    setFilters(updatedFilters);
  };

  const filtersActive = Object.values(filters).some(filter => !!filter.value);

  return (
    <FiltersContext.Provider value={{filters, applyFilters, filtersActive}}>
      {children}
    </FiltersContext.Provider>
  );
};
