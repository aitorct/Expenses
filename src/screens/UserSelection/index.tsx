import React, {useState} from 'react';
import {
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import {FlashList} from '@shopify/flash-list';
import {useSafeAreaInsets} from 'react-native-safe-area-context';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import strings from '@locales/index';
import colors from '@theme/colors';
import fontSizes from '@theme/fontSizes';
import sizes from '@theme/sizes';

const filterData = ['Aitor Cubeles Torres'];
const filterValue = filterData[0];

const UserSelection = () => {
  const [search, setSearch] = useState('');
  const insets = useSafeAreaInsets();

  const onUserSelect = () => {
    // Apply filter
  };

  const renderItem = ({item}: {item: string}) => {
    const isSelected = item === filterValue;

    const customStyle = isSelected
      ? styles.selectedItem
      : styles.unselectedItem;

    const onPress = () => {
      Keyboard.dismiss();
      onUserSelect();
    };

    return (
      <TouchableOpacity
        onPress={() => onPress()}
        style={[styles.item, customStyle]}>
        <Text style={styles.itemText}>{item}</Text>
        {item === filterValue && (
          <Icon name="check-circle" color={colors.icon} size={sizes.icon} />
        )}
      </TouchableOpacity>
    );
  };

  const filteredData = filterData.filter(item =>
    item.toLowerCase().includes(search.toLowerCase()),
  );
  const customListStyles = {paddingBottom: insets.bottom};

  return (
    <View style={[styles.container, styles.flex]}>
      <View style={styles.input}>
        <View style={styles.inputIcon}>
          <Icon name="magnify" color={colors.icon} size={sizes.icon} />
        </View>
        <TextInput
          style={styles.textInput}
          selectionColor={colors.selectionColor}
          placeholder={strings.filters.user.selection.placeholder}
          placeholderTextColor={colors.placeholderText}
          value={search}
          onChangeText={text => setSearch(text)}
        />
      </View>
      {filteredData.length ? (
        <KeyboardAvoidingView
          style={styles.flex}
          behavior="padding"
          keyboardVerticalOffset={sizes.s100}>
          <FlashList
            showsVerticalScrollIndicator={false}
            data={filteredData}
            contentContainerStyle={customListStyles}
            renderItem={renderItem}
            keyboardShouldPersistTaps="handled"
            estimatedItemSize={LIMIT}
          />
        </KeyboardAvoidingView>
      ) : (
        <View style={[styles.emptyState, styles.center]}>
          <Text style={styles.emptyText}>
            {strings.filters.user.selection.emptySearch}
          </Text>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    paddingTop: sizes.s16,
    paddingHorizontal: sizes.s16,
  },
  center: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  input: {
    marginTop: sizes.s8,
    marginBottom: sizes.s12,
    borderRadius: 12,
    paddingVertical: Platform.OS === 'ios' ? sizes.s8 : undefined,
    paddingHorizontal: sizes.s8,
    backgroundColor: colors.textInputBackground,
    alignItems: 'center',
    flexDirection: 'row',
  },
  inputIcon: {
    paddingRight: sizes.s8,
  },
  item: {
    height: sizes.s48,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: sizes.s16,
    marginBottom: sizes.s4,
    borderRadius: sizes.s8,
    borderWidth: sizes.border,
  },
  selectedItem: {
    borderColor: colors.selectedTouchableTagBorder,
    backgroundColor: colors.tagBackground,
  },
  unselectedItem: {
    borderColor: colors.touchableTagBackground,
    backgroundColor: colors.touchableTagBackground,
  },
  emptyState: {
    paddingTop: sizes.s24,
  },
  emptyText: {
    fontSize: fontSizes.largeText,
    textAlign: 'center',
  },
  textInput: {
    fontSize: fontSizes.text,
  },
  itemText: {
    fontSize: fontSizes.largeText,
    color: colors.text,
  },
});

export default UserSelection;
