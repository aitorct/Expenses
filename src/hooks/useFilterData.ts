import {ExpenseRealmContext} from '@contexts/ExpenseRealmContext';
import {FilterType} from '@contexts/FiltersContext/types';
import {Expense} from '@models/Expense';
import {ObjectExpenseData} from '@models/types';

const {useQuery} = ExpenseRealmContext;

type QueryableFilters = Exclude<FilterType, FilterType.RECEIPTS>;

type Queries = {
  [key in QueryableFilters]: {
    query: string;
    mapProperty: keyof ObjectExpenseData;
  };
};

const filterQueryTemplates: Queries = {
  year: {
    query: 'TRUEPREDICATE SORT(year DESC) DISTINCT(year)',
    mapProperty: 'year',
  },
  user: {
    query: 'TRUEPREDICATE SORT(user_name DESC) DISTINCT(user_name)',
    mapProperty: 'user_name',
  },
};

const useFilterData = (filterType: QueryableFilters) => {
  const results = useQuery(Expense);

  const {query, mapProperty} = filterQueryTemplates[filterType];

  const filteredResult = results
    .filtered(query)
    .map((item: ObjectExpenseData) => item[mapProperty].toString());

  return {filterData: filteredResult};
};

export default useFilterData;
