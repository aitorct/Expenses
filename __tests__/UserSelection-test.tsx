import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import {FiltersProvider} from '@contexts/FiltersContext';
import useFilterData from '@hooks/useFilterData';
import UserSelection from '@screens/UserSelection';

const mockedSafeAreaMetrics = {
  frame: {x: 0, y: 0, width: 0, height: 0},
  insets: {top: 0, left: 0, right: 0, bottom: 0},
};

const mockedUserName = 'Aitor Cubeles Torres';

jest.mock('@hooks/useFilterData', () =>
  jest.fn(() => ({
    filterData: [],
  })),
);

const mockUseFilterData = useFilterData as jest.Mock;

it('UserSelection with no data renders empty state correctly', () => {
  const tree = renderer
    .create(
      <SafeAreaProvider initialMetrics={mockedSafeAreaMetrics}>
        <FiltersProvider>
          <UserSelection />
        </FiltersProvider>
      </SafeAreaProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('UserSelection with data renders correctly', () => {
  mockUseFilterData.mockImplementation(
    jest.fn(() => ({
      filterData: [mockedUserName],
    })),
  );

  const tree = renderer
    .create(
      <SafeAreaProvider initialMetrics={mockedSafeAreaMetrics}>
        <FiltersProvider>
          <UserSelection />
        </FiltersProvider>
      </SafeAreaProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
