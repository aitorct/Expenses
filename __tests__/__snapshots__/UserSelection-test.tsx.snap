// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSelection with data renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 16,
          "paddingTop": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#7D828530",
          "borderRadius": 12,
          "flexDirection": "row",
          "marginBottom": 12,
          "marginTop": 8,
          "paddingHorizontal": 8,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          Object {
            "paddingRight": 8,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰍉
        </Text>
      </View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Type a user name..."
        placeholderTextColor="#7D8285"
        selectionColor="#000000"
        style={
          Object {
            "fontSize": 16,
          }
        }
        value=""
      />
    </View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RCTScrollView
          applyWindowCorrection={[Function]}
          canChangeSize={true}
          contentContainerStyle={
            Object {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
            }
          }
          contentHeight={0}
          contentWidth={0}
          data={
            Array [
              "Aitor Cubeles Torres",
            ]
          }
          dataProvider={
            DataProvider {
              "_data": Array [
                "Aitor Cubeles Torres",
              ],
              "_firstIndexToProcess": 0,
              "_hasStableIds": true,
              "_requiresDataChangeHandling": false,
              "_size": 1,
              "getStableId": [Function],
              "rowHasChanged": [Function],
            }
          }
          disableRecycling={false}
          estimatedItemSize={25}
          extendedState={Object {}}
          externalScrollView={[Function]}
          finalRenderAheadOffset={250}
          forceNonDeterministicRendering={true}
          horizontal={false}
          initialOffset={0}
          initialRenderIndex={0}
          isHorizontal={false}
          keyboardShouldPersistTaps="handled"
          layoutProvider={
            GridLayoutProviderWithProps {
              "_acceptableRelayoutDelta": 1,
              "_getHeightOrWidth": [Function],
              "_getLayoutTypeForIndex": [Function],
              "_getSpan": [Function],
              "_maxSpan": 1,
              "_setLayoutForType": [Function],
              "_tempDim": Object {
                "height": 0,
                "width": 0,
              },
              "averageWindow": AverageWindow {
                "currentAverage": 25,
                "currentCount": 1,
                "inputValues": Array [
                  25,
                ],
                "nextIndex": 1,
              },
              "defaultEstimatedItemSize": 100,
              "layoutObject": Object {
                "size": undefined,
                "span": undefined,
              },
              "props": Object {
                "contentContainerStyle": Object {
                  "paddingBottom": 0,
                },
                "data": Array [
                  "Aitor Cubeles Torres",
                ],
                "estimatedItemSize": 25,
                "keyboardShouldPersistTaps": "handled",
                "numColumns": 1,
                "renderItem": [Function],
                "showsVerticalScrollIndicator": false,
              },
              "shouldRefreshWithAnchoring": true,
            }
          }
          maxRenderAhead={750}
          numColumns={1}
          onEndReached={[Function]}
          onEndReachedThreshold={0}
          onEndReachedThresholdRelative={0}
          onItemLayout={[Function]}
          onLayout={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onSizeChanged={[Function]}
          onVisibleIndicesChanged={[Function]}
          removeClippedSubviews={false}
          renderAheadOffset={0}
          renderAheadStep={250}
          renderContentContainer={[Function]}
          renderItem={[Function]}
          renderItemContainer={[Function]}
          rowRenderer={[Function]}
          scrollEventThrottle={16}
          scrollThrottle={16}
          scrollViewProps={
            Object {
              "contentContainerStyle": Object {
                "backgroundColor": undefined,
                "minHeight": 1,
                "minWidth": 1,
              },
              "onLayout": [Function],
              "onScrollBeginDrag": [Function],
              "refreshControl": undefined,
              "style": Object {
                "minHeight": 1,
                "minWidth": 1,
              },
            }
          }
          showsVerticalScrollIndicator={false}
          style={
            Object {
              "minHeight": 1,
              "minWidth": 1,
            }
          }
          suppressBoundedSizeException={true}
          windowCorrectionConfig={
            Object {
              "applyToInitialOffset": true,
              "applyToItemScroll": true,
              "value": Object {
                "endCorrection": 0,
                "startCorrection": 0,
                "windowShift": -0,
              },
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <AutoLayoutView
                enableInstrumentation={false}
                horizontal={false}
                onBlankAreaEvent={[Function]}
                onLayout={[Function]}
                renderAheadOffset={0}
                scrollOffset={0}
                style={
                  Object {
                    "height": 0,
                    "width": 0,
                  }
                }
                windowSize={0}
              />
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`UserSelection with no data renders empty state correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 16,
          "paddingTop": 16,
        },
        Object {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#7D828530",
          "borderRadius": 12,
          "flexDirection": "row",
          "marginBottom": 12,
          "marginTop": 8,
          "paddingHorizontal": 8,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          Object {
            "paddingRight": 8,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰍉
        </Text>
      </View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Type a user name..."
        placeholderTextColor="#7D8285"
        selectionColor="#000000"
        style={
          Object {
            "fontSize": 16,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "paddingTop": 24,
          },
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "textAlign": "center",
          }
        }
      >
        No matching users found
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
